components/abstractions/src/main/java/com/microsoft/kiota/store/InMemoryBackingStore.java

```diff
@@ -212,39 +232,50 @@ private void setupNestedSubscriptions(
+        for (final Map.Entry<String, Pair<Boolean, Object>> entry : this.store.entrySet()) {
+            final Pair<Boolean, Object> wrapper = entry.getValue();
+            if (wrapper.getValue1() instanceof Pair) {
+                final Pair<?, Integer> collectionTuple = (Pair<?, Integer>) wrapper.getValue1();
+                Object[] items = getObjectArrayFromCollectionWrapper(collectionTuple);
+                for (Object item : items) {
+                    if (!(item instanceof BackedModel)) break;
+                    nestedBackedModelsToEnumerate.add((BackedModel) item);
+                }
+                if (collectionTuple.getValue1()
+                        != items.length) { // and the size has changed since we last updated
+                    currentStoreDirtyCollections.put(entry.getKey(), collectionTuple.getValue0());
+                }
             }
+        }
+
+        // Enumerate nested backed models first since they may trigger the parent to be dirty
+        for (BackedModel nestedBackedModel : nestedBackedModelsToEnumerate) {
+            nestedBackedModel.getBackingStore().enumerate();
+        }
 
-            if (collectionTuple.getValue1()
-                    != items.length) { // and the size has changed since we last updated
-                set(
-                        key,
-                        collectionTuple.getValue0()); // ensure the store is notified the collection
-                // property is "dirty"
+        // Only update parent properties that haven't been marked as dirty by the nested models
+        for (Map.Entry<String, Object> entry : currentStoreDirtyCollections.entrySet()) {
+            // Always set() if there were no nested models
+            if (nestedBackedModelsToEnumerate.isEmpty()) {
+                set(entry.getKey(), entry.getValue());
+                continue;
+            }
+            boolean hasChanged = this.store.get(entry.getKey()).getValue0();
+            if (!hasChanged) {
+                set(entry.getKey(), entry.getValue());
             }
         }
-        touchNestedProperties(storeItem); // call get on nested properties
     }
 
-    private void touchNestedProperties(final Object nestedObject) {
-        if (nestedObject instanceof BackedModel) {
-            // Call Get<>() on nested properties so that this method may be called recursively to
-            // ensure collections are consistent
-            final BackedModel backedModel = (BackedModel) nestedObject;
-            for (final String itemKey : backedModel.getBackingStore().enumerate().keySet()) {
-                backedModel.getBackingStore().get(itemKey);
-            }
+    private Object[] getObjectArrayFromCollectionWrapper(final Pair<?, Integer> collectionTuple) {
+        if (collectionTuple.getValue0() instanceof Collection) {
+            return ((Collection<Object>) collectionTuple.getValue0()).toArray();
+        } else { // it is a map
```

Comment:
could we test it so we don't end up throwing a cast exception for a case we didn't plan for?